docs/rfcs/024-extension-loading.md

```diff
@@ -0,0 +1,187 @@
+- When explicitly loading a library with `LOAD`
+- Wnen creating extension with `CREATE EXTENSION` (download sql scripts, (optional) extension data files and (optional) library files)))
+
+
+#### Summary
+
+Pros:
+ - Startup is only as slow as it takes to load all (shared_)preload_libraries
+ - Supports BYO Extension
+
+Cons:
+ - O(sizeof(extensions)) IO requirement for loading all extensions.
+
+### Alternative solutions
+
+1. Allow users to add their extensions to the base image
+   
+   Pros:
+    - Easy to deploy
+
+   Cons:
+    - Doesn't scale - first start size is dependent on image size;
+    - All extensions are shared across all users: It doesn't allow users to
+      bring their own restrictive-licensed extensions
+
+2. Bring Your Own compute image
+   
+   Pros:
+    - Still easy to deploy
+    - User can bring own patched version of PostgreSQL
+
+   Cons:
+    - First start latency is O(sizeof(extensions image))
+    - Warm instance pool for skipping pod schedule latency is not feasible with
+      O(n) custom images
+    - Support channels are difficult to manage
+
+3. Download all user extensions in bulk on compute start
+   
+   Pros:
+    - Easy to deploy
+    - No startup latency issues for "clean" users.
+    - Warm instance pool for skipping pod schedule latency is possible
+
+   Cons:
+    - Downloading all extensions in advance takes a lot of time, thus startup
+      latency issues
+
+4. Store user's extensions in persistent storage
+   
+   Pros:
+    - Easy to deploy
+    - No startup latency issues
+    - Warm instance pool for skipping pod schedule latency is possible
+
+   Cons:
+    - EC2 instances have only limited number of attachments shared between EBS
+      volumes, direct-attached NVMe drives, and ENIs.
+    - Compute instance migration isn't trivially solved for EBS mounts (e.g.
+      the device is unavailable whilst moving the mount between instances).
+    - EBS can only mount on one instance at a time (except the expensive IO2
+      device type).
+
+5. Store user's extensions in network drive
+   
+   Pros:
+    - Easy to deploy
+    - Few startup latency issues
+    - Warm instance pool for skipping pod schedule latency is possible
+
+   Cons:
+    - We'd need networked drives, and a lot of them, which would store many
+      duplicate extensions.
+    - **UNCHECKED:** Compute instance migration may not work nicely with
+      networked IOs
+
+
+### Idea extensions
+
+The extension store does not have to be S3 directly, but could be a Node-local
+caching service on top of S3. This would reduce the load on the network for
+popular extensions.
+
+## Extension Storage implementation
+
+The layout of the S3 bucket is as follows:
+```
+v14/ext_index.json
+    -- this contains information necessary to create control files
+v14/extensions/test_ext1.tar.gz
+    -- this contains the library files and sql files necessary to create this extension
+v14/extensions/custom_ext1.tar.gz
```

Comment:
How does extension versioning organized? 
In previous iteration we had `/<build-version>/` part of the path.

In this version how do we know that extension was build with compute v1234 and is compatible with it?

Reply:
cc @bayandin

Reply:
what if we did this:

```
build1234/ext1.tar.gz
build1234/ext2.tar.gz
build1235/ext1.tar.gz
build1236/ext2.tar.gz

build1234/ext_index.json
	- enable ext2 for tenant alek
	- enable ext1 for all tenants
build1235/ext_index.json
	- enable ext2 for tenant alek
	- enable ext1 for all tenants
build1236/ext_index.json
	- enable ext2 for tenant alek
	- enable ext1 for all tenants

extension_table.json contains the following data:
ext1:
	build1234: build1234
	build1235: build1234
	build1236: build1236
ext2:
	build1234: build1234
	build1235: build1235
	build1236: build1235
```

Then, when we download extensions the process is like this:
look in `build_number/ext_index.json` for a list of enabled extensions. 
When downloading an extension, go to the location specified in `extension_table.json` based on your build number. 

Basically, `extension_table.json` makes it so that we don't have to upload duplicates of extensions when they have not changed

Reply:
this is fixed now